### PR TITLE
Refactor RepoStatusWorker

### PR DESCRIPTION
**Pull request by Flipez**
Move from one big worker and data blob to repo assinged data

This fixes https://github.com/voxpupuli/vox-pupuli-tasks/issues/146 https://github.com/voxpupuli/vox-pupuli-tasks/issues/82

Signed-off-by: Flipez <code@brauser.io>

### CODE REVIEW

**Comment by Flipez** [comment]
Quick Note:
Use Database object instead of hash
Keep multiple objects per repo

This allows analysis and statistics

**Comment by bastelfreak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/models/repository.rb  14
         

Complexity decreasing per file
==============================
+ app/workers/repo_status_worker.rb  -17
         

Clones added
============
- app/models/repository.rb  3
         

Clones removed
==============
+ app/workers/repo_status_worker.rb  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834&bid=17854270)

**Comment by bastelfreak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- app/models/repository_status.rb  5
         

Complexity decreasing per file
==============================
+ app/workers/repo_status_worker.rb  -17
         

Clones removed
==============
+ app/workers/repo_status_worker.rb  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834&bid=17854270)

File: app/models/repository_status.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    if msync_file
+      self.never_synced = false
+      msync_version = YAML.safe_load(msync_file)['modulesync_config_version']
+
+      self.needs_another_sync = (
+        Gem::Version.new(RepoStatusData.latest_modulesync_version) >
+          Gem::Version.new(msync_version)
+      )
+    else
+      self.never_synced = false
```
Comment (bastelfreak) [inline app/models/repository_status.rb:40]: Codacy Issue found: [Move `self.never_synced = false` out of the conditional.](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: app/models/repository_status.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        support_range = "#{os_type.upcase}_SUPPORT_RANGE".constantize.all_i
+        supported_versions = os['operatingsystemrelease'].all_i
+      end
+
+      self.send(
+        "supports_eol_#{os_type.downcase}=",
+        supported_versions.min < support_range.min
+      )
+
+      self.send(
```
Comment (bastelfreak) [inline app/models/repository_status.rb:85]: Codacy Issue found: [Redundant `self` detected.](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

**Comment by bastelfreak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/models/repository_status.rb  5
         

Complexity decreasing per file
==============================
+ app/workers/repo_status_worker.rb  -17
         

Clones removed
==============
+ app/workers/repo_status_worker.rb  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834&bid=17854270)

File: app/lib/repository_check.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+class RepositoryCheck
```
Comment (bastelfreak) [inline app/lib/repository_check.rb:1]: Codacy Issue found: [Missing frozen string literal comment.](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

**Comment by bastelfreak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 10
           

Complexity increasing per file
==============================
- app/lib/repository_check.rb  2
- app/models/repository_status.rb  5
         

Complexity decreasing per file
==============================
+ app/workers/repo_status_worker.rb  -17
         

Clones removed
==============
+ app/workers/repo_status_worker.rb  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834&bid=17854270)

File: config/initializers/voxpupuli.rb [hunk 4]
```diff
@@ -11,3 +11,7 @@
 CENTOS_SUPPORT_RANGE  = VOXPUPULI_CONFIG.dig('support_ranges', 'centos')
 FREEBSD_SUPPORT_RANGE = VOXPUPULI_CONFIG.dig('support_ranges', 'freebsd')
 FEDORA_SUPPORT_RANGE  = VOXPUPULI_CONFIG.dig('support_ranges', 'fedora')
+
+REPOSITORY_STATUS_CHECKS = []
```
Comment (bastelfreak) [inline config/initializers/voxpupuli.rb:15]: Codacy Issue found: [Freeze mutable objects assigned to constants.](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: app/lib/repository_check.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  def self.init
+    REPOSITORY_STATUS_CHECKS << name
+  end
+
+  def self.load_checks
+    Dir[File.join("#{Rails.root}/app/lib/repository_checks", "*.rb")].each do |file|
```
Comment (bastelfreak) [inline app/lib/repository_check.rb:7]: Codacy Issue found: [Prefer single-quoted strings when you don't need string interpolation or special symbols.](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: app/lib/repository_checks/missing_in_plumbing.rb [hunk 6]
```diff
@@ -0,0 +1,4 @@
+class MissingInPlumbing < RepositoryCheck
+  def perform(repository)
+  end
+end
```
Comment (bastelfreak) [inline app/lib/repository_checks/missing_in_plumbing.rb:4]: Codacy Issue found: [Final newline missing.](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: config/initializers/voxpupuli.rb [hunk 7]
```diff
@@ -11,3 +11,7 @@
...
+
+RepositoryCheck.load_checks
```
Comment (bastelfreak) [inline config/initializers/voxpupuli.rb:17]: Codacy Issue found: [Final newline missing.](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: app/lib/repository_check.rb (hunk 5 above)
Comment (bastelfreak) [inline app/lib/repository_check.rb:7]: Codacy Issue found: [Sort files before requiring them.](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: app/lib/repository_checks/missing_in_plumbing.rb [hunk 8]
```diff
@@ -0,0 +1,4 @@
+class MissingInPlumbing < RepositoryCheck
+  def perform(repository)
```
Comment (bastelfreak) [inline app/lib/repository_checks/missing_in_plumbing.rb:2]: Codacy Issue found: [Put empty method definitions on a single line.](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: app/lib/repository_checks/missing_in_plumbing.rb [hunk 9]
```diff
@@ -0,0 +1,4 @@
+class MissingInPlumbing < RepositoryCheck
```
Comment (bastelfreak) [inline app/lib/repository_checks/missing_in_plumbing.rb:1]: Codacy Issue found: [Missing frozen string literal comment.](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: app/lib/repository_check.rb (hunk 5 above)
Comment (bastelfreak) [inline app/lib/repository_check.rb:7]: Codacy Issue found: [Please use `Rails.root.join('path/to')` instead.](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: app/lib/repository_check.rb [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+      next unless valid?(clazz)
+
+      clazz.init
+    end
+  end
+
+  def self.valid?(clazz)
+    clazz.method_defined?('perform') && (clazz < self)
+  end
+end
```
Comment (bastelfreak) [inline app/lib/repository_check.rb:20]: Codacy Issue found: [Final newline missing.](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

**Comment by bastelfreak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- app/lib/repository_check.rb  2
- app/models/repository_status.rb  5
         

Complexity decreasing per file
==============================
+ app/workers/repo_status_worker.rb  -17
         

Clones removed
==============
+ app/workers/repo_status_worker.rb  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834&bid=17854270)

File: config/initializers/voxpupuli.rb (hunk 4 above)
Comment (bastelfreak) [inline config/initializers/voxpupuli.rb:15]: Codacy Issue found: [Freeze mutable objects assigned to constants.](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: config/initializers/voxpupuli.rb (hunk 7 above)
Comment (bastelfreak) [inline config/initializers/voxpupuli.rb:17]: Codacy Issue found: [Final newline missing.](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

**Comment by bastelfreak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/lib/repository_check_base.rb  2
- app/lib/repository_checks/need_initial_modulesync.rb  2
- app/lib/repository_checks/need_initial_release.rb  2
- app/lib/repository_checks/metadata.rb  4
- app/lib/repository_checks/modulesync_file.rb  2
         

Complexity decreasing per file
==============================
+ app/workers/repo_status_worker.rb  -17
         

Clones removed
==============
+ app/workers/repo_status_worker.rb  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834&bid=17854270)

**Comment by bastelfreak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/lib/repository_checks/need_initial_modulesync.rb  2
- app/lib/repository_checks/modulesync_file.rb  2
- app/lib/repository_check_base.rb  2
- app/lib/repository_checks/metadata.rb  3
- app/lib/repository_checks/need_initial_release.rb  2
         

Complexity decreasing per file
==============================
+ app/workers/repo_status_worker.rb  -17
         

Clones removed
==============
+ app/workers/repo_status_worker.rb  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834&bid=17854270)

**Comment by bastelfreak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- app/lib/repository_checks/metadata.rb  3
- app/lib/repository_check_base.rb  2
         

Complexity decreasing per file
==============================
+ app/workers/repo_status_worker.rb  -17
         

Clones removed
==============
+ app/workers/repo_status_worker.rb  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834&bid=17854270)

File: app/lib/repository_checks/missing_secrets.rb [hunk 11]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class MissingSecrets < RepositoryCheckBase
+  def perform
+    submit_result :has_secrets = Github.get_file(repo.full_name, '.sync.yml')
```
Comment (bastelfreak) [inline app/lib/repository_checks/missing_secrets.rb:5]: Codacy Issue found: [unexpected token tEQL
(Using Ruby 2.6 parser; configure using `TargetRubyVersion` parameter, under `AllCops`)](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: app/lib/repository_checks/need_initial_release.rb [hunk 12]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+class NeedInitialRelease < RepositoryCheckBase
+  def perform
+    submit_result :released = (RepoStatusData.forge_releases.include?(repo.name) ||
```
Comment (bastelfreak) [inline app/lib/repository_checks/need_initial_release.rb:5]: Codacy Issue found: [unexpected token tEQL
(Using Ruby 2.6 parser; configure using `TargetRubyVersion` parameter, under `AllCops`)](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: app/lib/repository_checks/without_reference_dot_md.rb [hunk 13]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class WithoutReferenceDotMd < RepositoryCheckBase
+  def perform
+    submit_result :reference_dot_md = Github.get_file(repo.full_name, 'REFERENCE.md')
```
Comment (bastelfreak) [inline app/lib/repository_checks/without_reference_dot_md.rb:5]: Codacy Issue found: [unexpected token tEQL
(Using Ruby 2.6 parser; configure using `TargetRubyVersion` parameter, under `AllCops`)](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: app/lib/repository_checks/need_initial_modulesync.rb [hunk 14]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+class NeedInitialModulesync < RepositoryCheckBase
+  def perform
+    submit_result :has_modulesync = (RepoStatusData.modulesync_repos.include?(repo.name) ||
```
Comment (bastelfreak) [inline app/lib/repository_checks/need_initial_modulesync.rb:5]: Codacy Issue found: [unexpected token tEQL
(Using Ruby 2.6 parser; configure using `TargetRubyVersion` parameter, under `AllCops`)](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: app/lib/repository_checks/missing_in_modulesync_repo.rb [hunk 15]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class MissingInModulesyncRepo < RepositoryCheckBase
+  def perform
+    submit_result :in_modulesync_repo = RepoStatusData.modulesync_repos.include?(repo.name)
```
Comment (bastelfreak) [inline app/lib/repository_checks/missing_in_modulesync_repo.rb:5]: Codacy Issue found: [unexpected token tEQL
(Using Ruby 2.6 parser; configure using `TargetRubyVersion` parameter, under `AllCops`)](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: app/lib/repository_checks/modulesync_file.rb [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+
+class ModulesyncFile < RepositoryCheckBase
+  def perform
+    msync_file = Github.get_file(repo.full_name, '.msync.yml')
+
+    if msync_file
+      submit_result :synced, true
+      msync_version = YAML.safe_load(msync_file)['modulesync_config_version']
+
+      submit_result :latest_modulesync = (
```
Comment (bastelfreak) [inline app/lib/repository_checks/modulesync_file.rb:11]: Codacy Issue found: [unexpected token tEQL
(Using Ruby 2.6 parser; configure using `TargetRubyVersion` parameter, under `AllCops`)](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: app/lib/repository_checks/missing_in_plumbing.rb [hunk 17]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class MissingInPlumbing < RepositoryCheckBase
+  def perform
+    submit_result :in_plumbing = RepoStatusData.plumbing_modules.include?(repo.name)
```
Comment (bastelfreak) [inline app/lib/repository_checks/missing_in_plumbing.rb:5]: Codacy Issue found: [unexpected token tEQL
(Using Ruby 2.6 parser; configure using `TargetRubyVersion` parameter, under `AllCops`)](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: app/lib/github.rb [hunk 18]
```diff
@@ -15,6 +15,16 @@ def self.client
...
   end
 
+  def self.get_file(repo, path)
+    client.contents(repo, path: path, headers: { 'Accept': 'application/vnd.github.v3.raw' })
+  rescue Octokit::NotFound
+    false
+  end
+
+  def self.check_file(repo, path)
+    !!get_file(repo, path)
```
Comment (bastelfreak) [inline app/lib/github.rb:25]: Codacy Issue found: [Avoid the use of double negation (`!!`).](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

**Comment by bastelfreak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- app/lib/repository_checks/need_initial_release.rb  2
- app/lib/repository_checks/metadata.rb  3
- app/lib/repository_checks/modulesync_file.rb  2
- app/lib/repository_check_base.rb  2
- app/lib/repository_checks/need_initial_modulesync.rb  2
         

Complexity decreasing per file
==============================
+ app/workers/repo_status_worker.rb  -17
         

Clones removed
==============
+ app/workers/repo_status_worker.rb  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834&bid=17854270)

File: app/lib/github.rb (hunk 18 above)
Comment (bastelfreak) [inline app/lib/github.rb:25]: Codacy Issue found: [Avoid the use of double negation (`!!`).](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

Comment (bastelfreak) [inline app/lib/github.rb:25]: Codacy Issue found: [Avoid the use of double negation (`!!`).](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

**Comment by bastelfreak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- app/lib/repository_checks/need_initial_release.rb  2
- app/lib/repository_checks/need_initial_modulesync.rb  2
- app/lib/repository_checks/modulesync_file.rb  2
- app/lib/repository_check_base.rb  2
- app/lib/repository_checks/metadata.rb  3
         

Complexity decreasing per file
==============================
+ app/workers/repo_status_worker.rb  -17
         

Clones removed
==============
+ app/workers/repo_status_worker.rb  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834&bid=17854270)

**Comment by bastelfreak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- app/lib/repository_checks/modulesync_file.rb  2
- app/lib/repository_check_base.rb  2
- app/lib/repository_checks/need_initial_release.rb  2
- app/lib/repository_checks/metadata.rb  3
- app/lib/repository_checks/need_initial_modulesync.rb  2
         

Complexity decreasing per file
==============================
+ app/workers/repo_status_worker.rb  -17
         

Clones removed
==============
+ app/workers/repo_status_worker.rb  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834&bid=17854270)

File: app/lib/github.rb (hunk 18 above)
Comment (bastelfreak) [inline app/lib/github.rb:25]: Codacy Issue found: [Avoid the use of double negation (`!!`).](https://app.codacy.com/manual/VoxPupuli/vox-pupuli-tasks/pullRequest?prid=5444834)

File: app/controllers/checks_controller.rb [hunk 19]
```diff
@@ -0,0 +1,6 @@
+class ChecksController < ApplicationController
+  def show
+    @repositories = Repository.all
+    @check_name = params[:check_name]
+  end
+end
```
Comment (bastelfreak) [inline app/controllers/checks_controller.rb:6]: please fix the missing \n in line 6

File: app/models/pull_request.rb [hunk 20]
```diff
@@ -167,6 +167,9 @@ def validate(saved_changes)
     # Don't run through validation if it's a draft
     return if draft
 
+    # Don't run through validation if it's a draft
+    return if draft
```
Comment (bastelfreak) [inline app/models/pull_request.rb:171]: you doubled the line?

File: app/views/checks/show.html.haml [hunk 21]
```diff
@@ -0,0 +1,10 @@
+.ui.dividing.header= @check_name.titleize
+.ui.relaxed.divided.list
+  - @repositories.each do |repo|
+    = link_to(repository_path(repo.name), class: :item) do
+      = repo.name
+      .right.floated
+        - if repo.passed_check?(@check_name)
+          .ui.basic.horizontal.green.label pass
+        - else
+          .ui.basic.horizontal.red.label miss
```
Comment (bastelfreak) [inline app/views/checks/show.html.haml:10]: please add missing `\n`

File: app/views/dashboard/_check_card.haml [hunk 22]
```diff
@@ -0,0 +1,13 @@
...
+      = check_title
+      %span{data: { tooltip: I18n.t("check_descriptions.#{check_name}"), position: 'top center'}}
+        %i.icon.question.mark.circle.outline.tiny
+  .content
+    - pass_count = @statuses.count{|rs| rs.checks[check_name]}
+    %div{style: 'padding-bottom: 10px;'}
+    %span.ui.huge.text{class: color_by_passed_checks(pass_count, @statuses.count)}= pass_count
+    %span.ui.big.grey.text{style: 'padding-left: 5px;'}
+      \/
+      = @statuses.count
```
Comment (bastelfreak) [inline app/views/dashboard/_check_card.haml:13]: please add missing `\n`

File: app/views/dashboard/show.html.haml [hunk 23]
```diff
@@ -1,32 +1,53 @@
...
+  = render partial: 'check_card', locals: {check_title: 'Latest msync', check_name: 'latest_modulesync'}
+
+.ui.large.header Dependencies
+.ui.four.cards
+  = render partial: 'check_card', locals: {check_title: 'Has PVR', check_name: 'with_puppet_version_range'}
+  = render partial: 'check_card', locals: {check_title: 'Correct PVR', check_name: 'correct_puppet_version_range'}
+  = render partial: 'check_card', locals: {check_title: 'Has Reference.md', check_name: 'reference_dot_md'}
+  = render partial: 'check_card', locals: {check_title: 'Has secrets', check_name: 'has_secrets'}
+
+.ui.basic.segment= area_chart PullRequest.group_by_day(:gh_created_at).count, title: 'Pull requests per day', library: {chart: {zoomType: 'x'}}
```
Comment (bastelfreak) [inline app/views/dashboard/show.html.haml:53]: please add missing `\n`

**Pull request merged**